reps/2023-12-14-virtual-cluster.md

```diff
@@ -0,0 +1,452 @@
+
+### Should this change be within `ray` or outside?
+
+Inside `ray` project since this is a Ray Core feature.
+
+## Stewardship
+
+### Required Reviewers
+
+@ericl, @stephanie-wang, @scv119
+
+### Shepherd of the Proposal
+
+@ericl
+
+## Design
+
+With the introduction of virtual clusters, every Ray job runs in its own virtual cluster and only has access to (logical) resources inside that virtual cluster. Each virtual cluster has a spec that defines the min and max resources of the cluster. Min resources are minimal resources required for the job to run and they are atomically reserved for gang scheduling. If min resources cannot be reserved when there are not enough available resources, the job will be queued. With job queueing, we can implement different policies such as FIFO or priority-based queueing. Max resources are the autoscaling limit of the virtual cluster and the maximal resources can be used by the job.
+
+Virtual clusters can be nested and a Ray job can create sub-clusters to isolate separate parts of its application workload. For example, a Tune grid sweep job can create a sub-cluster for each of its nested Train workload. These possibly nested virtual clusters form a tree where the root is the entire Ray cluster.
+
+<img src="https://user-images.githubusercontent.com/898023/291139618-0be11470-db09-466d-8c2c-37b9e5b3765c.png" alt="drawing" width="508"/>
+
+Virtual clusters with different min and max resources are autoscalable. When scaling up, virtual clusters will try to borrow more resources from their parent clusters. If their parents have available resources, the scaling up is instant. Otherwise, parents will try to borrow resources from their parents recursively and eventually this may cause the upscale of the physical cluster. When scaling down, virtual clusters will return resources to their parents recursively and eventually this may cause the downscale of the physical cluster.
+
+Virtual nodes of a virtual cluster can be either fixed size or flexible/resizable. For a single virtual cluster, there can be multiple fixed-size virtual nodes but at most one flexible virtual node on a single physical node. The upscaling of a virtual cluster can be achieved by adding new virtual nodes or scaling up an existing flexible virtual node.
+
+### API
+
+#### Virtual Cluster Spec
+
+```
+message VirtualCluster {
+  // A virtual cluster consits of flexible resources and fixed size resources.
+
+  // == Flexible resources ==
+  // Flexible resources are stored in flexible virtual nodes
+  // and Ray will use as few flexible virtual nodes as possible
+  // for the given amount of flexible resources to minimize fragmentation.
+  // In other words, Ray prefers to scale up a flexible virtual node
+  // than adding a new flexible virtual node.
+
+  // If specified, ensure we have at least this min amount
+  // of resources before starting the cluster.
+  // If not specified, the default value is 0.
+  map<string, double> flexible_resource_min
```

Comment:
```suggestion
  map<string, double> flex_resource_min
```

nit: "flexible" is too long/hard to spell, prefer to abbreviate in these cases

Reply:
should we just name it total_resource_min, total_resource_max?

Reply:
> should we just name it min_resources, max_resources?

The semantics is different. `flex_resource_min` doesn't include fixed node resources while `min_resources` should include everything.